pkg/docker/docker_test.go

```diff
@@ -24,201 +30,336 @@ func TestContainerName(t *testing.T) {
 }
 
 func getDocker(client Client) *stiDocker {
+	//k8s has its own fake docker client mechanism
+	k8sDocker := dockertools.ConnectToDockerOrDie("fake://")
 	return &stiDocker{
-		client:   client,
-		pullAuth: docker.AuthConfiguration{},
+		kubeDockerClient: k8sDocker,
+		client:           client,
+		pullAuth:         dockertypes.AuthConfig{},
 	}
 }
 
+//NOTE, neither k8s kube client nor engine api make their error types public, so we can access / instantiate them like we did for the analogous
+// ones in go-dockerclient.
+// Hence, we test with errors created in these tests and set on the fake client  vs. those actually defined in k8s/engine-api.
+// Also, the fake client does save the parameters passed in; but instead, you can confirm that it was called via AssertCalls
+
 func TestIsImageInLocalRegistry(t *testing.T) {
 	type testDef struct {
-		imageName         string
-		docker            test.FakeDockerClient
-		expectedImageName string
-		expectedResult    bool
-		expectedError     error
+		imageName      string
+		docker         test.FakeDockerClient
+		expectedResult bool
+		expectedError  string
 	}
-	otherError := fmt.Errorf("Other")
 	tests := map[string]testDef{
-		"ImageFound":    {"a_test_image", test.FakeDockerClient{Image: &docker.Image{}}, "a_test_image:latest", true, nil},
-		"ImageNotFound": {"a_test_image:sometag", test.FakeDockerClient{InspectImageErr: []error{docker.ErrNoSuchImage}}, "a_test_image:sometag", false, nil},
-		"ErrorOccurred": {"a_test_image", test.FakeDockerClient{InspectImageErr: []error{otherError}}, "a_test_image:latest", false, otherError},
+		"ImageFound":    {"a_test_image", test.FakeDockerClient{}, true, ""},
+		"ImageNotFound": {"a_test_image:sometag", test.FakeDockerClient{}, false, "unable to get metadata for a_test_image:sometag"},
 	}
 
 	for test, def := range tests {
 		dh := getDocker(&def.docker)
+		fake := dh.kubeDockerClient.(*dockertools.FakeDockerClient)
+		if def.expectedResult {
+			fake.Image = &dockertypes.ImageInspect{ID: def.imageName}
+		}
+
 		result, err := dh.IsImageInLocalRegistry(def.imageName)
+
+		if e := fake.AssertCalls([]string{"inspect_image"}); e != nil {
+			t.Errorf("%+v", e)
+		}
+
 		if result != def.expectedResult {
 			t.Errorf("Test - %s: Expected result: %v. Got: %v", test, def.expectedResult, result)
 		}
-		if err != def.expectedError {
-			t.Errorf("Test - %s: Expected error: %v. Got: %v", test, def.expectedError, err)
-		}
-		if def.docker.InspectImageName[0] != def.expectedImageName {
-			t.Errorf("Docker inspect called with unexpected image name: %s",
-				def.docker.InspectImageName)
+		if err != nil && len(def.expectedError) > 0 && !strings.Contains(err.Error(), def.expectedError) {
+			t.Errorf("Test - %s: Expected error: Got: %+v", test, err)
 		}
 	}
 }
-
 func TestCheckAndPullImage(t *testing.T) {
+	// in addition to the deltas mentioned up top, kube fake client can't distinguish between exists and successful pull, and
```

Comment:
There is an image field that is publicly accessible that we can manipulate to discern whether the image exists locally:  `Image                *dockertypes.ImageInspect`

And for pull, you can either inject an error, or add a image name to a list of pulled images which is *NOT* publicly accessible, either by itself or from a public method:  `pulled               []string`